evm/src/cpu/kernel/asm/curve_add.asm

```diff
@@ -0,0 +1,383 @@
+    DUP2
+    // stack: y0, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x0, y0, x0, y0, x1, y1, retdest
+    %ec_check
+    // stack: isValid(x0, y0), x0, y0, x1, y1, retdest
+    DUP5
+    // stack: x1, isValid(x0, y0), x0, y0, x1, y1, retdest
+    DUP5
+    // stack: x1, y1, isValid(x0, y0), x0, y0, x1, y1, retdest
+    %ec_check
+    // stack: isValid(x1, y1), isValid(x0, y0), x0, y0, x1, y1, retdest
+    AND
+    // stack: isValid(x1, y1) & isValid(x0, y0), x0, y0, x1, y1, retdest
+    PUSH ec_add_valid_points
+    // stack: ec_add_valid_points, isValid(x1, y1) & isValid(x0, y0), x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    POP
+    // stack: y0, x1, y1, retdest
+    POP
+    // stack: x1, y1, retdest
+    POP
+    // stack: y1, retdest
+    POP
+    // stack: retdest
+    %ec_invalid_input
+
+// Assumption: (x0,y0) and (x1,y1) are valid points.
+global ec_add_valid_points:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+    DUP2
+    // stack: y0, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x0, y0, x0, y0, x1, y1, retdest
+    %ec_isidentity
+    // stack: (x0,y0)==(0,0), x0, y0, x1, y1, retdest
+    PUSH ec_add_first_zero
+    // stack: ec_add_first_zero, y0==0 & x0==0, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    DUP4
+    // stack: y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: x1, y1, x0, y0, x1, y1, retdest
+    %ec_isidentity
+    // stack: (x1,y1)==(0,0), x0, y0, x1, y1, retdest
+    PUSH ec_add_snd_zero
+    // stack: ec_add_snd_zero, y1==0 & x1==0, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    DUP3
+    // stack: x1, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x0, x1, x0, y0, x1, y1, retdest
+    EQ
+    // stack: x0 == x1, x0, y0, x1, y1, retdest
+    PUSH ec_add_equal_first_coord
+    // stack: ec_add_equal_first_coord, x0 == x1, x0, y0, x1, y1, retdest
+    JUMPI
+    // stack: x0, y0, x1, y1, retdest
+    DUP4
+    // stack: y1, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: y0, y1, x0, y0, x1, y1, retdest
+    %submod
+    // stack: y0 - y1, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: x1, y0 - y1, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: x0, x1, y0 - y1, x0, y0, x1, y1, retdest
+    %submod
+    // stack: x0 - x1, y0 - y1, x0, y0, x1, y1, retdest
+    %moddiv
+    // stack: lambda, x0, y0, x1, y1, retdest
+    PUSH ec_add_valid_points_with_lambda
+    // stack: ec_add_valid_points_with_lambda, lambda, x0, y0, x1, y1, retdest
+    JUMP
+
+// Assumption: (x0,y0) == (0,0)
+ec_add_first_zero:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+    POP
+    // stack: y0, x1, y1, retdest
+    POP
+    // stack: x1, y1, retdest
+    DUP2
+    // stack: y1, x1, y1, retdest
+    DUP2
+    // stack: x1, y1, x1, y1, retdest
+    %ec_isidentity
+    // stack: (x1,y1)==(0,0), x1, y1, retdest
+    PUSH ret_zero
+    // stack: ret_zero, (x1,y1)==(0,0), x1, y1, retdest
+    JUMPI
```

Comment:
Do we need to handle this separately? Seems like the general case of returning (x1, y1), below, would still work if it's the identity?